docs/Test.md

```diff
@@ -0,0 +1,154 @@
+for example) or the end user of the UI.
+
+`Test driven development` tests conversely target individual functions in a
+more traditional unit test style - validating that for a given input, the
+expected output is returned.
+
+Both styles of test should live in the same directory as the components they
+are testing.
+
+Depending on the code being covered, the tools used to enable these test styles
+may be different, as detailed below.
+
+### Test Driven Development
+
+Code in the following directories are expected to be driven by standard test
+driven development. They use _Jest_ as a test runner, using standard Jest
+test assets, such as `describe`, `it` and `expect`:
+
+- `client/Hooks`
+- `client/Contexts`
+
+### Behavioural Driven Development
+
+Code in the following directories are expected to be driven by behavioural
+driven deveopment. They use _Jest_ as a test runner, using standard Jest test
+assets, such as `describe`, `it` and `expect`, and _RTL_ to mount and interact
+with React component code:
+
+- `client/Elements`
+- `client/Groups`
+- `client/Panels`
+
+### End to End testing
+
+End to End tests should exercise a user scenario - focus on the behaviour a
+user will take to interact with the application to achieve an end goal.
+
+They should be fairly coarse grained, exercising multiple pages/panels as
+necessary.
+
+For example - "I can navigate to the topics page and see a list topics", or "I
+can navigate to a topic and view its consumers".
+
+These tests are using _cypress_ as a test runner and to drive assertions.
+Cypress tests should be written like how a user would interact with the browser
+when using the UI.
+
+To avoid the temptation to correlate scenarios with particular panels, these
+tests are located in the `e2e/features` directory.
+
+The e2e tests are written using [Gerkin](https://cucumber.io/docs/gherkin/reference)
+syntax. Your `.feature` file should live inside `e2e/features`. The steps are
+written inside `e2e/features/step_definitions`. Definition files are written using
+[cypress-cucumber-preprocessor](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor).
+
+On failure - the `failure_output/screenshots` and `failure_output/videos`
+directories will contain a capture of the page at failure time, and a
+recordings of the test until failure.
+
+## Writing a test
+
+In all styles of test, we will look for the following factors. These are to
+confirm the test is clean and will be reliable:
+
+- Date and time should be mocked/set to a known value
+- Tests should cause no side effects - they should create and delete any
+  resources they need in the scope of that test/test suite
+- Tests (regardless of style) should be structured in the following way:
+  1. Perform any required setup
+  2. Perform an action (or actions)
+  3. Assert the result
+  4. Perform any required clean up
+
+## Code coverage
+
+In this repo, we aim to have 100% code coverage across all components. This code
```

Comment:
I had an idea around coverage: based on the files changed, run just related tests via Jest (for example), and get coverage on just those results. Ie, tests related to the change made to check coverage - no chance of having another test happen to 'run over' lines.

I can see this being non trivial, but will mean an area can be fully excercised by it's tests alone.

Reply:
For now, I'll just expand on what we believe the use of code coverage is. Further investigation on selective test coverage is needed before this can be fleshed out. I have raised an investigation point on the project board https://github.com/orgs/strimzi/projects/2#card-46309994.